pkg/provisioner/ironic/ironic.go

```diff
@@ -715,6 +715,158 @@ func (p *ironicProvisioner) getUpdateOptsForNode(ironicNode *nodes.Node) (update
+	logicalDisks, err := p.buildTargetRAIDCfg()
+	if len(logicalDisks) == 0 || err != nil {
+		return
+	}
+	// Set target for RAID configuration steps
+	err = nodes.SetRAIDConfig(
+		p.client,
+		ironicNode.UUID,
+		nodes.RAIDConfigOpts{LogicalDisks: logicalDisks},
+	).ExtractErr()
+	if err != nil {
+		p.log.Error(err, "failed in setting the target RAID configuration",
+			"logical disks", logicalDisks,
+		)
+	}
+	return
+}
+
+// A private method to build the clean steps for RAID configuration from BaremetalHost spec
+func (p *ironicProvisioner) buildRAIDCleanSteps() (cleanSteps []nodes.CleanStep) {
+	// To add ‘delete_configuration’ before ‘create_configuration’ to make sure
+	cleanSteps = append(
+		cleanSteps,
+		nodes.CleanStep{
+			Interface: "raid",
+			Step:      "delete_configuration",
+		},
+	)
+	// If not configure raid, only need to clear old configuration
+	if p.host.Spec.RAID == nil || p.host.Spec.RAID.Volumes == nil || len(p.host.Spec.RAID.Volumes) == 0 {
+		return
+	}
+	if p.host.Spec.RAID.IsSoftwareRaid {
+		cleanSteps = append(
+			cleanSteps,
+			nodes.CleanStep{
+				Interface: "deploy",
+				Step:      "erase_devices_metadata",
+			},
+		)
+	}
+	// ‘create_configuration’ doesn’t remove existing disks. It is recommended
+	// that only the desired logical disks exist in the system after manual cleaning.
+	cleanSteps = append(
+		cleanSteps,
+		nodes.CleanStep{
+			Interface: "raid",
+			Step:      "create_configuration",
+		},
+	)
+	return
+}
+
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	cleanSteps := make([]nodes.CleanStep, 0)
+
+	if p.host.Spec.RAID != nil && p.host.Spec.RAID.DeleteExisting {
+		// Set raid config
+		err = p.setTargetRAIDCfg(ironicNode)
+		if err != nil {
+			return
+		}
+		// Build raid configuration clean steps
+		cleanSteps = append(cleanSteps, p.buildRAIDCleanSteps()...)
+	}
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		return p.changeNodeProvisionState(ironicNode, opts)
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.Manageable:
+		result, err = p.startManualCleaning(ironicNode)
+		if ironicNode, err = p.findExistingHost(); err != nil {
+			return result, errors.Wrap(err, "could not find host to clean")
+		}
+		if len(ironicNode.CleanStep) != 0 {
+			result.Dirty = false
+		}
```

Comment:
This is a good thought, but it would probably be more reliable if we provide a way for callers of `changeNodeProvisionState()` to determine if we're delaying because something actually happened or because the host was locked, as in your previous implementation. I just proposed #536 to make that simple.

Reply:
Thanks. After #536 is merged, I will rebase this PR.